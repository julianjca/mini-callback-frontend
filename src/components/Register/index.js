import React, { useState } from 'react'
import axios from 'axios'
import {
  Center,
  Container,
  Heading,
  Stack,
  Input,
  FormControl,
  FormLabel,
  Button,
} from '@chakra-ui/react'

const Register = ({ setRenderRegisterPage }) => {
  const [name, setName] = useState('')
  const [password, setPassword] = useState('')
  const [email, setEmail] = useState('')

  const handleSubmit = async () => {
    await axios.post('http://localhost:3030/users', {
      name,
      password,
      email,
    })
    setRenderRegisterPage(false)
  }

  return (
    <Container>
      <Center>
        <Heading>Register</Heading>
      </Center>
      <Center mt={10}>
        <Stack spacing={5}>
          <FormControl id="name">
            <FormLabel>Name</FormLabel>
            <Input
              value={name}
              width={300}
              type="text"
              placeholder="John"
              size="md"
              onChange={e => setName(e.target.value)}
              data-testid="name-input"
            />
          </FormControl>
          <FormControl id="email">
            <FormLabel>Email address</FormLabel>
            <Input
              width={300}
              type="email"
              placeholder="john@mail.com"
              size="md"
              onChange={e => setEmail(e.target.value)}
              data-testid="email-input"
            />
          </FormControl>
          <FormControl id="password">
            <FormLabel>Password</FormLabel>
            <Input
              width={300}
              type="password"
              placeholder="Enter your password"
              size="md"
              onChange={e => setPassword(e.target.value)}
              data-testid="password-input"
            />
          </FormControl>
          <Button onClick={handleSubmit} size="md" width={300} colorScheme="blue">
            Sign up
          </Button>
        </Stack>
      </Center>
    </Container>
  )
}

export default Register
