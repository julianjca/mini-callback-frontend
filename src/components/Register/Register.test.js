/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react'
import { render, fireEvent, waitFor } from '../../../test-utils'
import axios from 'axios'

import Register from './index'

jest.mock('axios')

describe('Register', () => {
  test('renders the right text', () => {
    const element = render(<Register />)

    expect(element).toMatchSnapshot()
  })

  test('renders email input, name input, and password input', () => {
    const element = render(<Register />)

    expect(element.getByTestId('email-input')).toBeInTheDocument()
    expect(element.getByTestId('password-input')).toBeInTheDocument()
    expect(element.getByTestId('name-input')).toBeInTheDocument()
  })

  test('runs register function', async () => {
    const setRenderRegisterPage = jest.fn()
    const element = render(<Register setRenderRegisterPage={setRenderRegisterPage} />)

    const axiosMock = jest.spyOn(axios, 'post').mockImplementation(() => Promise.resolve())

    const emailInput = element.getByTestId('email-input')
    const passwordInput = element.getByTestId('password-input')
    const nameInput = element.getByTestId('name-input')

    fireEvent.change(emailInput, { target: { value: 'john@mail.com' } })
    fireEvent.change(passwordInput, { target: { value: 'password' } })
    fireEvent.change(nameInput, { target: { value: 'John' } })

    const button = element.getByText('Sign up')
    fireEvent.click(button)

    await waitFor(() => expect(axiosMock).toHaveBeenCalledTimes(1))
    await waitFor(() =>
      expect(axiosMock).toHaveBeenCalledWith('http://localhost:3030/users', {
        email: 'john@mail.com',
        password: 'password',
        name: 'John',
      }),
    )

    await waitFor(() => expect(setRenderRegisterPage).toHaveBeenCalledTimes(1))
    await waitFor(() => expect(setRenderRegisterPage).toHaveBeenCalledWith(false))

    axiosMock.mockRestore()
    setRenderRegisterPage.mockReset()
  })
})
