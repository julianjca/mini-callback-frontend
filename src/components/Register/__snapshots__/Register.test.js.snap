// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders the right text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="chakra-container css-1plwrig"
      >
        <div
          class="css-gmuwbf"
        >
          <h2
            class="chakra-heading css-zey6tx"
          >
            Register
          </h2>
        </div>
        <div
          class="css-rwokqt"
        >
          <div
            class="chakra-stack css-ynqzo5"
          >
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="name"
                id="name-label"
              >
                Name
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="name-input"
                id="name"
                placeholder="John"
                type="text"
                value=""
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="email"
                id="email-label"
              >
                Email address
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="email-input"
                id="email"
                placeholder="john@mail.com"
                type="email"
              />
            </div>
            <div
              class="chakra-form-control css-1kxonj9"
              role="group"
            >
              <label
                class="chakra-form__label css-1qqx8g3"
                for="password"
                id="password-label"
              >
                Password
              </label>
              <input
                class="chakra-input css-1hv7q3p"
                data-testid="password-input"
                id="password"
                placeholder="Enter your password"
                type="password"
              />
            </div>
            <button
              class="chakra-button css-pa74a0"
              type="button"
            >
              Sign up
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-container css-1plwrig"
    >
      <div
        class="css-gmuwbf"
      >
        <h2
          class="chakra-heading css-zey6tx"
        >
          Register
        </h2>
      </div>
      <div
        class="css-rwokqt"
      >
        <div
          class="chakra-stack css-ynqzo5"
        >
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="name"
              id="name-label"
            >
              Name
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="name-input"
              id="name"
              placeholder="John"
              type="text"
              value=""
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="email"
              id="email-label"
            >
              Email address
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="email-input"
              id="email"
              placeholder="john@mail.com"
              type="email"
            />
          </div>
          <div
            class="chakra-form-control css-1kxonj9"
            role="group"
          >
            <label
              class="chakra-form__label css-1qqx8g3"
              for="password"
              id="password-label"
            >
              Password
            </label>
            <input
              class="chakra-input css-1hv7q3p"
              data-testid="password-input"
              id="password"
              placeholder="Enter your password"
              type="password"
            />
          </div>
          <button
            class="chakra-button css-pa74a0"
            type="button"
          >
            Sign up
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
